src/PreconditionerTypes.f90

```diff
@@ -343,6 +550,695 @@ SUBROUTINE setup_ILU_PreCondtype(thisPC)
     ENDIF
   ENDIF
 ENDSUBROUTINE setup_ILU_PreCondtype
+!
+!-------------------------------------------------------------------------------
+!> @brief Intializes up the SOR Preconditioner Type with a parameter list
+!> @param thisPC The preconditioner to act on
+!> @param A The matrix to precondition
+!> @param params The parameter list
+!>
+SUBROUTINE init_SOR_PreCondtype(thisPC,A,params)
+  CHARACTER(LEN=*),PARAMETER :: myName='init_RSOR_PreCondType'
+  CLASS(SOR_PrecondType),INTENT(INOUT) :: thisPC
+  CLASS(MatrixType),ALLOCATABLE,TARGET,INTENT(IN),OPTIONAL :: A
+  TYPE(ParamType),INTENT(IN),OPTIONAL :: params
+  TYPE(ParamType)::PListMat_LU
+  INTEGER(SIK)::k
+
+  !required statuses
+  REQUIRE(.NOT. thisPC%isinit)
+  REQUIRE(PRESENT(A))
+  REQUIRE(ALLOCATED(A))
+  REQUIRE(A%isInit)
+
+  thisPC%A => A
+
+  !gets the number of blocks from the parameter list
+  CALL params%get('PCType->numBlocks',thisPC%numBlocks)
+  CALL params%get('PCType->omega',thisPC%omega)
+
+  !makes sure that the number of blocks is valid
+  REQUIRE(thisPC%numBlocks .GT. 0)
+
+  !makes sure that the number of blocks is valid
+  REQUIRE(MOD(thisPC%A%n,thisPC%numBlocks) .EQ. 0)
+
+  !makes sure that the value of omega is valid
+  REQUIRE(thisPC%omega .LE. 2)
+  REQUIRE(thisPC%omega .GE. 0)
+
+  !calculate block size
+  thisPC%blockSize=thisPC%A%n/thisPC%numBlocks
+
+  !makes a lu matrix for each diagonal block in an array
+  ALLOCATE(DenseSquareMatrixType :: thisPC%LU(thisPC%numBlocks))
+
+  !initializes those matrices
+  CALL PListMat_LU%add('MatrixType->n',thisPC%blockSize)
+  CALL PListMat_LU%add('MatrixType->isSym',.FALSE.)
+  DO k=1,thisPC%numBlocks
+    CALL thisPC%LU(k)%init(PListMat_LU)
+  ENDDO
+
+  SELECT TYPE(mat => thisPC%A)
+  TYPE IS(DenseSquareMatrixType)
+    ALLOCATE(DenseSquareMatrixType :: thisPC%LpU)
+    ! Assign A to LpU
+    SELECT TYPE(LpU => thisPC%LpU); TYPE IS(DenseSquareMatrixType)
+      LpU=mat
+    ENDSELECT
+    REQUIRE(thisPC%LpU%isInit)
```

Comment:
If at the top `REQUIRE(A%isInit)`, `thisPC%A => A`, so `%A` is init, and we assign mat to `%thisPC%A`, and then we assign mat to LpU, requiring `LpU%isInit` is redundant.  We can get rid of all of them.

Reply:
This is correct for the serial types. In all cases, the serial types require a duplication of the matrix (without the dense diagonal blocks) to work. The initialization sets LpU (the duplicated matrix) to A because of this. The setup routine then goes through LpU and makes the necessary changes, which cannot be made to A without loss of information. So LpU itself has to stay.

This would be expensive and not worth our time except for the fact that the blockbanded matrix is designed in a specific way to avoid this copying. In that case, LpU is set as a pointer => A because A will have a routine to "mask" the dense blocks without copying.

I will get rid of the extraneous require statements though.